### PR TITLE
Morphological ranked freetext search over the node text and answers

### PR DESCRIPTION
**Pull request by crdev**

### CODE REVIEW

File: bot.py [hunk 1]
```diff
@@ -273,6 +279,24 @@ def choice(update: Update, context: CallbackContext) -> int:
 
     if update.message.text in CONVERSATION_DATA["node_by_name"]:
         next_node_name = update.message.text
+    elif not update.message.text in CUSTOM_NODES:
```
Comment (skd) [inline bot.py:282]: Please move this logic to ConversationHandler filters.

File: morpho_index.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import pymorphy2
+import re
+
+from multiset import Multiset
+from node_util import visit_node
+from operator import itemgetter
+from typing import Dict, List, Set, Tuple
+
+SPLIT_REGEX = re.compile("[^а-яА-Яa-zA-Z-]+")
+RUSSIAN_WORD = re.compile("^[а-яА-Я-]$")
```
Comment (skd) [inline morpho_index.py:12]: What about hyphens?

File: morpho_index.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+			for item in node_counts.items():
+				count = found_word_count_by_node_name.get(item[0], 0)
+				found_word_count_by_node_name[item[0]] = count + 1
+			for item in node_counts.items():
+				result_multiset.add(item[0], item[1])
+
+		# Boost nodes having hits for multiple words from the query.
+		nodes_and_scores = [(node_name, count * found_word_count_by_node_name[node_name]) for (node_name, count) in result_multiset.items()]
+		nodes_and_scores.sort(key=itemgetter(1), reverse=True)
+		logger.info(f"Search: [{text}] -> {nodes_and_scores}")
```
Comment (skd) [inline morpho_index.py:104]: .debug

File: bot.py [hunk 4]
```diff
@@ -273,6 +279,24 @@ def choice(update: Update, context: CallbackContext) -> int:
...
+        search_results = morpho_index.search(update.message.text)
+        if search_results:
+            next_node_name = search_results[0][0]
+        else:
+            if FEEDBACK_CHANNEL_ID is not None:
+                context.bot.send_message(
```
[the remark this answers is not in the document]
Comment (crdev) [reply bot.py:288]: Done

File: bot.py [hunk 5]
```diff
@@ -273,6 +279,24 @@ def choice(update: Update, context: CallbackContext) -> int:
 
     if update.message.text in CONVERSATION_DATA["node_by_name"]:
         next_node_name = update.message.text
+    elif not update.message.text in CUSTOM_NODES:
+        search_results = morpho_index.search(update.message.text)
+        if search_results:
+            next_node_name = search_results[0][0]
```
[the remark this answers is not in the document]
Comment (crdev) [reply bot.py:285]: Done

File: bot.py (hunk 1 above)
Comment (crdev) [reply bot.py:282]: That won't help. As discussed in chat, added an "organic_call" default param to choice() to distinguish between the CHOICE state vs. others.

File: bot.py [hunk 6]
```diff
@@ -295,21 +319,8 @@ def choice(update: Update, context: CallbackContext) -> int:
...
-        current_keyboard_options.append([START_OVER])
-    else:
-        if FEEDBACK_CHANNEL_ID is not None:
-            current_keyboard_options.append([FEEDBACK])
-        if update.message.from_user.username in ADMIN_USERS:
-            current_keyboard_options.appendleft([ADMIN])
-
-    current_keyboard = ReplyKeyboardMarkup(
-        current_keyboard_options, one_time_keyboard=True)
+    logger.info(f"{current_node_name} - {next_node_name}")
```
[the remark this answers is not in the document]
Comment (crdev) [reply bot.py:322]: Sorry, a leftover, removed.

File: bot.py [hunk 7]
```diff
@@ -329,6 +340,24 @@ def choice(update: Update, context: CallbackContext) -> int:
     return CHOOSING
 
 
+def build_keyboard_options(current_node_name: str, next_node_name: str, update: Update, user_data: Dict[str, str]):
```
[the remark this answers is not in the document]
Comment (crdev) [reply bot.py:343]: Could you perhaps take another look?

File: stats.py [hunk 8]
```diff
@@ -32,7 +32,7 @@ def store_interaction(self, user_id: str, node: str, ts: int):
     def get_users_data(self) -> List[int]:
         pass
 
-    def get_interactions_data(self) -> Counter[str]:
+    def get_interactions_data(self) -> Counter:
```
[the remark this answers is not in the document]
Comment (crdev) [reply stats.py:35]: Done

File: morpho_index.py (hunk 3 above)
Comment (crdev) [reply morpho_index.py:104]: Done.

File: morpho_index.py (hunk 2 above)
Comment (crdev) [reply morpho_index.py:12]: They are there! :) See the last char in each char class.

File: bot.py [hunk 9]
```diff
@@ -273,6 +280,21 @@ def choice(update: Update, context: CallbackContext) -> int:
 
     if update.message.text in CONVERSATION_DATA["node_by_name"]:
         next_node_name = update.message.text
+    elif organic_call:
+        search_results = morpho_index.search(update.message.text)
+        if search_results:
+            next_node_name = search_results[0][0]
+            update.message.reply_text(SEARCH_RESULT_HEADER_TEMPLATE.format(next_node_name))
```
Comment (skd) [inline bot.py:287]: Maybe replace {} with % and call SEARCH_RESULT_HEADER_TEMPLATE % next_node_name?

File: bot.py [hunk 10]
```diff
@@ -329,6 +338,25 @@ def choice(update: Update, context: CallbackContext) -> int:
     return CHOOSING
 
 
+def build_keyboard_options(keyboard_options_node: str, update: Update, nav_stack_depth: int):
```
Comment (skd) [inline bot.py:341]: Nit: can you replace update with show_admin_button?

File: morpho_index.py (hunk 2 above)
Comment (skd) [reply morpho_index.py:12]: It seems that I'm blind.

File: bot.py (hunk 10 above)
Comment (crdev) [reply bot.py:341]: Done.

File: bot.py (hunk 9 above)
Comment (crdev) [reply bot.py:287]: As discussed, format() is the modern way of string formatting.

File: morpho_index.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+				wt = word_tag(word)
+				if not wt:
+					continue
+				node_set = self.node_counts_by_word_tag.setdefault(wt, Multiset())
+				node_set.add(node.name, weight)
+
+		def process_node(node: conversation_proto.ConversationNode):
+			if node.name in IGNORED_NODES:
+				return
+			process_text(node, node.name, 5)
```
Comment (skd) [inline morpho_index.py:74]: Make the score 9000 [1]. :) I think.

[1] https://en.wikipedia.org/wiki/It%27s_Over_9000!

**Review by skd — APPROVED** [review]

**Pull request merged**